src/stream/src/executor/backfill/snapshot_backfill/vnode_stream.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+use std::mem::{replace, take};
+use std::pin::Pin;
+use std::task::{ready, Context, Poll};
+
+use futures::stream::{FuturesUnordered, Peekable, StreamFuture};
+use futures::{Stream, StreamExt, TryStreamExt};
+use risingwave_common::array::{Op, StreamChunk};
+use risingwave_common::hash::VirtualNode;
+use risingwave_common::row::{OwnedRow, Row, RowExt};
+use risingwave_common::util::chunk_coalesce::DataChunkBuilder;
+
+use crate::executor::StreamExecutorResult;
+
+pub(super) type BackfillRowItem = ((Op, OwnedRow), Option<(Op, OwnedRow)>);
+pub(super) trait BackfillRowStream =
+    Stream<Item = StreamExecutorResult<BackfillRowItem>> + Sized + 'static;
+
+struct StreamWithVnode<St> {
+    stream: St,
+    vnode: VirtualNode,
+}
+
+impl<St: Stream + Unpin> Stream for StreamWithVnode<St> {
+    type Item = St::Item;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.stream.poll_next_unpin(cx)
+    }
+}
+
```

Comment:
Can we adapt the docs from the PR and add them here? To my understanding the reason we need a custom stream, rather than rely on a `try_stream` macro, is because we need to encapsulate some state inside the stream as well.

It is possible to do it with `try_stream`, but the internal state of the sink will be exposed.

Reply:
It's hard to use the `try_stream` macro, because the ownership of the vnode streams will be owned by the stream generated by `try_stream`, and then we can't get the state of the vnode streams.

Reply:
Okay let's add some docs to provide this justification.